egs/librispeech/asr/simple_v1/mmi_xent_train.py

```diff
@@ -0,0 +1,734 @@
+    parser.add_argument(
+        '--start-epoch',
+        type=int,
+        default=0,
+        help="Number of start epoch.")
+    parser.add_argument(
+        '--max-frames',
+        type=int,
+        default=90000,
+        help="Maximum number of feature frames in a single batch.") 
+    parser.add_argument(
+        '--den-scale',
+        type=float,
+        default=1.0,
+        help="denominator scale in mmi loss.")
+    parser.add_argument(
+        '--xent-scale',
+        type=float,
+        default=0.1,
+        help="Cross entropy loss scale.")
+    parser.add_argument(
+        '--smooth-scale',
+        type=float,
+        default=0.1,
+        help="Phone distribution smoothing loss scale.")
+    return parser
+
+
+def main():
+    args = get_parser().parse_args()
+
+    start_epoch = args.start_epoch
+    num_epochs = args.num_epochs
+    max_frames = args.max_frames
+    den_scale = args.den_scale
+    xent_scale = args.xent_scale
+    smooth_scale = args.smooth_scale
+
+    fix_random_seed(42)
+
+    exp_dir = Path('exp-lstm-adam-mmi-xent-musan-' + 'xent-scale-' + str(xent_scale) + '-smooth-scale-' + str(smooth_scale) )
+    setup_logger('{}/log/log-train'.format(exp_dir))
+    tb_writer = SummaryWriter(log_dir=f'{exp_dir}/tensorboard')
+
+    # load L, G, symbol_table
+    lang_dir = Path('data/lang_nosp')
+    phone_symbol_table = k2.SymbolTable.from_file(lang_dir / 'phones.txt')
+    word_symbol_table = k2.SymbolTable.from_file(lang_dir / 'words.txt')
+
+    logging.info("Loading L.fst")
+    if (lang_dir / 'Linv.pt').exists():
+        L_inv = k2.Fsa.from_dict(torch.load(lang_dir / 'Linv.pt'))
+    else:
+        with open(lang_dir / 'L.fst.txt') as f:
+            L = k2.Fsa.from_openfst(f.read(), acceptor=False)
+            L_inv = k2.arc_sort(L.invert_())
+            torch.save(L_inv.as_dict(), lang_dir / 'Linv.pt')
+
+    graph_compiler = MmiTrainingGraphCompiler(
+        L_inv=L_inv,
+        phones=phone_symbol_table,
+        words=word_symbol_table
+    )
+    phone_ids = get_phone_symbols(phone_symbol_table)
+    P = create_bigram_phone_lm(phone_ids)
+    P.scores = torch.zeros_like(P.scores)
+
+    # load dataset
+    feature_dir = Path('exp/data')
+    logging.info("About to get train cuts")
+    cuts_train = CutSet.from_json(feature_dir /
+                                  'cuts_train-clean-100.json.gz')
+    logging.info("About to get dev cuts")
+    cuts_dev = CutSet.from_json(feature_dir / 'cuts_dev-clean.json.gz')
+    logging.info("About to get Musan cuts")
+    cuts_musan = CutSet.from_json(feature_dir / 'cuts_musan.json.gz')
+
+    logging.info("About to create train dataset")
+    train = K2SpeechRecognitionDataset(
+        cuts_train,
+        cut_transforms=[
+            CutConcatenate(),
+            CutMix(
+                cuts=cuts_musan,
+                prob=0.5,
+                snr=(10, 20)
+            )
+        ]
+    )
+    train_sampler = SingleCutSampler(
```

Comment:
@danpovey since he is using a `SingleCutSampler` and `CutConcatenate` a couple of lines before, I don't believe there are any significant changes to how the batches are formed w.r.t. the past experiments. For these to go into effect, you'd have to use the `BucketingSampler`.

Reply:
Oh OK. 
BTW, RE BucketingSampler: another concern I have is that if we use that, we'd have to do the same in test time because the model won't have "seen" data where there is a discontinuity.

Reply:
I think that cut concatenation has always been disabled in test time (dev/test Dataset/Dataloader)